File: scripts/download_images.py
"""
Script download ảnh từ Joomla enews.agu.edu.vn về storage local.
Chạy từ thư mục gốc Laravel: python scripts/download_images.py --cookie "sl-session=OA1OPYuNw2l9XVATzPbNbw=="
"""
import argparse
import os
import sys
import time
import urllib.request
import sqlite3
import pymysql
import json

# ── Cấu hình DB (đọc từ .env) ──────────────────────────────────────────────
def read_env(env_path='.env'):
    config = {}
    with open(env_path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if '=' in line and not line.startswith('#'):
                k, _, v = line.partition('=')
                config[k.strip()] = v.strip().strip('"').strip("'")
    return config

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--cookie', required=True, help='sl-session cookie value (e.g. sl-session=abc123==)')
    parser.add_argument('--limit', type=int, default=0, help='Limit số ảnh (0=tất cả)')
    args = parser.parse_args()

    env = read_env('.env')
    db_host = env.get('DB_HOST', '127.0.0.1')
    db_port = int(env.get('DB_PORT', 3306))
    db_name = env.get('DB_DATABASE', 'enews')
    db_user = env.get('DB_USERNAME', 'root')
    db_pass = env.get('DB_PASSWORD', '')

    print(f"DB: {db_user}@{db_host}:{db_port}/{db_name}")

    try:
        conn = pymysql.connect(host=db_host, port=db_port, db=db_name,
                               user=db_user, password=db_pass, charset='utf8mb4')
        cursor = conn.cursor()
    except Exception as e:
        print(f"❌ Kết nối DB thất bại: {e}")
        sys.exit(1)

    # Lấy posts có thumbnail Joomla
    query = "SELECT id, thumbnail FROM posts WHERE thumbnail IS NOT NULL AND thumbnail != '' AND thumbnail NOT LIKE 'http%' AND thumbnail LIKE 'images/%'"
    if args.limit > 0:
        query += f" LIMIT {args.limit}"
    cursor.execute(query)
    posts = cursor.fetchall()
    total = len(posts)
    print(f"📊 Tổng số bài cần tải: {total}")

    # Thư mục lưu ảnh
    storage_dir = os.path.join('storage', 'app', 'public', 'joomla-images')
    os.makedirs(storage_dir, exist_ok=True)

    cookie = args.cookie  # ví dụ: "sl-session=OA1OPYuNw2l9XVATzPbNbw=="
    base_url = 'https://enews.agu.edu.vn'

    success = fail = skip = 0

    for i, (post_id, thumbnail) in enumerate(posts):
        remote_path = thumbnail.lstrip('/')
        remote_url  = f"{base_url}/{remote_path}"
        local_name  = os.path.basename(remote_path)
        local_path  = os.path.join(storage_dir, local_name)
        db_path     = f"joomla-images/{local_name}"

        # In progress
        pct = int((i+1)/total*100)
        print(f"\r[{i+1}/{total}] {pct}% — {local_name[:40]}", end='', flush=True)

        # Skip nếu đã tồn tại
        if os.path.exists(local_path) and os.path.getsize(local_path) > 100:
            cursor.execute("UPDATE posts SET thumbnail=%s WHERE id=%s", (db_path, post_id))
            conn.commit()
            skip += 1
            continue

        # Download
        try:
            req = urllib.request.Request(
                remote_url,
                headers={
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/122.0.0.0 Safari/537.36',
                    'Accept': 'image/avif,image/webp,image/apng,image/*,*/*;q=0.8',
                    'Referer': 'https://enews.agu.edu.vn/',
                    'Cookie': cookie,
                }
            )
            with urllib.request.urlopen(req, timeout=15) as resp:
                content = resp.read()

            if len(content) < 100:
                fail += 1
                continue

            with open(local_path, 'wb') as f:
                f.write(content)

            cursor.execute("UPDATE posts SET thumbnail=%s WHERE id=%s", (db_path, post_id))
            conn.commit()
            success += 1

        except Exception as e:
            fail += 1

    print(f"\n\n✅ Thành công : {success}")
    print(f"⏭️  Đã tồn tại : {skip}")
    print(f"❌ Thất bại   : {fail}")

    cursor.close()
    conn.close()

if __name__ == '__main__':
    main()

File: scripts/download-images.js
/**
 * download-images.js
 * Dùng Puppeteer (browser thật) để bypass WAF SafeLine và download ảnh Joomla
 * Usage: node scripts/download-images.js
 */

const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');
const mysql = require('mysql2/promise');

// Đọc .env
function readEnv() {
  const env = {};
  const lines = fs.readFileSync(path.join(__dirname, '..', '.env'), 'utf8').split('\n');
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#') && trimmed.includes('=')) {
      const [k, ...vs] = trimmed.split('=');
      env[k.trim()] = vs.join('=').trim().replace(/^["']|["']$/g, '');
    }
  }
  return env;
}

async function main() {
  const env = readEnv();
  const storageDir = path.join(__dirname, '..', 'storage', 'app', 'public', 'joomla-images');
  fs.mkdirSync(storageDir, { recursive: true });

  // Kết nối DB
  const db = await mysql.createConnection({
    host: env.DB_HOST || '127.0.0.1',
    port: parseInt(env.DB_PORT || '3306'),
    database: env.DB_DATABASE,
    user: env.DB_USERNAME,
    password: env.DB_PASSWORD,
  });
  console.log('DB connected:', env.DB_DATABASE);

  // Lấy danh sách ảnh cần tải
  const [rows] = await db.query(
    "SELECT id, thumbnail FROM posts WHERE thumbnail IS NOT NULL AND thumbnail != '' AND thumbnail NOT LIKE 'http%' AND thumbnail LIKE 'images/%'"
  );
  console.log(`Total images: ${rows.length}`);

  // Mở browser thật (bypass WAF)
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
  });
  const page = await browser.newPage();

  // Truy cập trang chủ Joomla trước để lấy session cookie (pass WAF)
  console.log('Accessing Joomla to get WAF session...');
  await page.goto('https://enews.agu.edu.vn', { waitUntil: 'networkidle2', timeout: 30000 });
  console.log('WAF passed! Starting downloads...');

  let success = 0, fail = 0, skip = 0;

  for (let i = 0; i < rows.length; i++) {
    const { id, thumbnail } = rows[i];
    const remotePath = thumbnail.replace(/^\//, '');
    const remoteUrl  = `https://enews.agu.edu.vn/${remotePath}`;
    const localName  = path.basename(remotePath);
    const localPath  = path.join(storageDir, localName);
    const dbPath     = `joomla-images/${localName}`;

    process.stdout.write(`\r[${i+1}/${rows.length}] ${Math.round((i+1)/rows.length*100)}% - ${localName.slice(0,40)}`);

    // Skip nếu đã có file
    if (fs.existsSync(localPath) && fs.statSync(localPath).size > 100) {
      await db.query('UPDATE posts SET thumbnail=? WHERE id=?', [dbPath, id]);
      skip++;
      continue;
    }

    try {
      // Download qua Puppeteer (có WAF cookies)
      const response = await page.goto(remoteUrl, { waitUntil: 'load', timeout: 15000 });
      if (!response || response.status() >= 400) { fail++; continue; }

      const buffer = await response.buffer();
      if (buffer.length < 100) { fail++; continue; }

      fs.writeFileSync(localPath, buffer);
      await db.query('UPDATE posts SET thumbnail=? WHERE id=?', [dbPath, id]);
      success++;
    } catch (e) {
      fail++;
    }
  }

  await browser.close();
  await db.end();

  console.log(`\n\nSuccess : ${success}`);
  console.log(`Skipped : ${skip}`);
  console.log(`Failed  : ${fail}`);
}

main().catch(err => { console.error(err); process.exit(1); });
